src/opera/template/topology.py

```diff
@@ -68,3 +68,117 @@ def get_artifact(self, params):
         node_name, *rest = params
 
         return self.find_node(node_name).get_artifact(["SELF"] + rest)
+
+    def concat(self, params):
+        if not isinstance(params, list):
+            raise DataError("Concat intrinsic function parameters for '{}'"
+                            " should be a list".format(params))
+
+        concatenated_string = ""
+        for param in params:
+            if isinstance(param, dict):
+                param_dict_keys = list(param.keys())
+                if len(param_dict_keys) > 1:
+                    raise DataError(
+                        "Dict value expression for concat function "
+                        "has too many keys: '{}'".format(param))
+
+                value_expression_name = param_dict_keys[0]
+                try:
+                    concatenated_string += getattr(self,
+                                                   value_expression_name)(
+                        param[value_expression_name])
+                except Exception:
+                    raise DataError(
+                        "Invalid value expression for concat "
+                        "function: '{}'".format(value_expression_name))
+
+            elif isinstance(param, str):
+                concatenated_string += param
+            else:
+                raise DataError(
+                    "Invalid value for concat function: '{}'".format(param))
+        return concatenated_string
+
+    def join(self, params):
```

Comment:
If **join** function with 1 parameter is the same as **concat** I would get rid of **concat** implementation and do something like 
```python
    def concat(self, params):
        if not isinstance(params, list):
            raise DataError("Concat intrinsic function parameters for '{}'"
                            " should be a list".format(params))

        return self.join([params])
```

and for **join**
```python
delimiter = "" if len(params) == 1 else params[1]
```

And for all errors use something like 
```
for concat or join function
```

Reply:
Thanks, good suggestion! I've refactored this part.